src/mscorlib/src/System/Threading/ThreadPool.cs

```diff
@@ -1625,7 +1640,16 @@ private static bool QueueUserWorkItemHelper(WaitCallback callBack, Object state,
                 try { }
                 finally
                 {
-                    QueueUserWorkItemCallback tpcallBack = new QueueUserWorkItemCallback(callBack, state, compressStack, ref stackMark);
+                    ExecutionContext context = compressStack && !ExecutionContext.IsFlowSuppressed() ?
+                                                ExecutionContext.Capture(ref stackMark, ExecutionContext.CaptureOptions.IgnoreSyncCtx | ExecutionContext.CaptureOptions.OptimizeDefaultCase) :
+                                                ExecutionContext.Default;
+
+                    QueueUserWorkItemCallback tpcallBack = (context == ExecutionContext.Default) ?
+                                                (state == null ? 
+                                                    new QueueUserWorkItemCallback(callBack) : // default or null context, no state 
+                                                    new QueueUserWorkItemCallbackWithState(callBack, state)) : // default or null context
```

Comment:
> So ExecutionContext Capture(ref stackMark, captureOptions) will return null on CoreCLR? If it doesn't then it will use the larger version?

Let's start over.  The optimization should be this:
- When ExecutionContext.Default is provided, create a queued work item object that doesn't store any context, and then when it's executed, uses ```ExecutionContext.Run(ExecutionContext.Default, ...).
- When any other ExecutinContext is provided, create a queued work item that stores the context (which may or may not be null), and does the same logic that's done today: if it's null, just invoke the callback, and if it's not null, use ```ExecuteContext.Run(context, ...)``` to run the callback.